tools/monitoring/mysql/alert.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysql
+
+import (
+	"database/sql"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/knative/test-infra/tools/monitoring/config"
+	"github.com/knative/test-infra/tools/monitoring/mail"
+)
+
+const (
+	alertInsertStmt = `
+	INSERT INTO Alerts (
+		ErrorPattern, Sent
+	) VALUES (?,?)`
+
+	emailTemplate = `In the past %v, 
+The number of occurrences of the following error pattern reached threshold:
+%s
+
+Hint for diagnose & recovery: %s
+`
+)
+
+func sendAlert(errorPattern string, config *config.SelectedConfig, mailConfig *mail.Config, recipients []string) error {
+	log.Printf("sending alert...")
+	subject := fmt.Sprintf("Error pattern reached alerting threshold: %s", errorPattern)
+	body := fmt.Sprintf(emailTemplate, config.Duration(), errorPattern, config.Hint)
+
+	return mailConfig.Send(recipients, subject, body)
+}
+
+// CheckAlertCondition checks whether the given error pattern meets
+// the alert condition specified in config
+func CheckAlertCondition(errorPattern string, config *config.SelectedConfig, db *sql.DB) (bool, error) {
+	// the timestamp we want to start collecting logs
+	startTime := time.Now().Add(config.Duration())
+
+	var nMatches, nJobs, nPRs int
+
+	row := db.QueryRow(`
+		SELECT 
```

Comment:
move to const?

Reply:
Actually I don't think we should always put queries in constants, especially for the ones that is only associated with a single function. It makes reading harder, because you have to look at two places to understand what is going on. Furthermore, these queries are not really constant, they have variables (?), it is hard to see what variable has what value when we have them in two places